test/framework/finalizers_helpers.go

```diff
@@ -0,0 +1,168 @@
+	removeFinalizers(ctx, proxy, namespace)
+
+	// Unpause the cluster.
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, false)
+
+	// Annotate the ClusterClass, if one is in use, to speed up reconciliation. This ensures ClusterClass Finalizers
+	// are re-reconciled before asserting the owner reference graph.
+	forceClusterClassReconcile(ctx, proxy.GetClient(), clusterKey)
+
+	// Annotate the MachineDeployment  to speed up reconciliation. This ensures MachineDeployment topology Finalizers are re-reconciled.
+	// TODO: Remove this as part of https://github.com/kubernetes-sigs/cluster-api/issues/9532
+	forceMachineDeploymentTopologyReconcile(ctx, proxy.GetClient(), clusterKey)
+
+	// Check that the Finalizers are as expected after further reconciliations.
+	assertFinalizersExist(ctx, proxy, namespace, objectsWithFinalizers)
+}
+
+// removeFinalizers removes all Finalizers from objects in the owner graph.
+func removeFinalizers(ctx context.Context, proxy ClusterProxy, namespace string) {
+	graph, err := clusterctlcluster.GetOwnerGraph(ctx, namespace, proxy.GetKubeconfigPath())
+	Expect(err).ToNot(HaveOccurred())
+	for _, object := range graph {
+		ref := object.Object
+		obj := new(unstructured.Unstructured)
+		obj.SetAPIVersion(ref.APIVersion)
+		obj.SetKind(ref.Kind)
+		obj.SetName(ref.Name)
+
+		Expect(proxy.GetClient().Get(ctx, client.ObjectKey{Namespace: namespace, Name: object.Object.Name}, obj)).To(Succeed())
+		helper, err := patch.NewHelper(obj, proxy.GetClient())
+		Expect(err).ToNot(HaveOccurred())
+		obj.SetFinalizers([]string{})
+		Expect(helper.Patch(ctx, obj)).To(Succeed())
+	}
+}
+
+func getObjectsWithFinalizers(ctx context.Context, proxy ClusterProxy, namespace string) map[string]*unstructured.Unstructured {
+	graph, err := clusterctlcluster.GetOwnerGraph(ctx, namespace, proxy.GetKubeconfigPath())
+	Expect(err).ToNot(HaveOccurred())
+
+	objsWithFinalizers := map[string]*unstructured.Unstructured{}
+
+	for _, node := range graph {
+		nodeNamespacedName := client.ObjectKey{Namespace: node.Object.Namespace, Name: node.Object.Name}
+		obj := &unstructured.Unstructured{}
+		obj.SetAPIVersion(node.Object.APIVersion)
+		obj.SetKind(node.Object.Kind)
+		err = proxy.GetClient().Get(ctx, nodeNamespacedName, obj)
+		Expect(err).ToNot(HaveOccurred())
+
+		setFinalizers := obj.GetFinalizers()
+
+		if len(setFinalizers) > 0 {
+			objsWithFinalizers[client.ObjectKey{Namespace: node.Object.Namespace, Name: node.Object.Name}.String()] = obj
+		}
+	}
+
+	return objsWithFinalizers
+}
+
+// assertFinalizersExist ensures that current Finalizers match those in the initialObjectsWithFinalizers.
+func assertFinalizersExist(ctx context.Context, proxy ClusterProxy, namespace string, initialObjsWithFinalizers map[string]*unstructured.Unstructured) {
+	Eventually(func() error {
+		var allErrs []error
+		finalObjsWithFinalizers := getObjectsWithFinalizers(ctx, proxy, namespace)
```

Comment:
Is it intentional that we only validate if the finalizers of objects with finalizers are correct?

E.g. this means we are missing if an expected finalizers is not set at all. I assume we need this in some way because the topology finalizers are not set in the legacy cluster case?

Reply:
Right, this was done for 2 reasons - first like you said because of legacy cluster cases and then the goal of this test to check if previously set finalisers were resilient to removal.

Reply:
Sound good enough!

Reply:
It does speed-up reconciliation for topology Clusters https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/topology/cluster/cluster_controller.go#L101